File: app/src/main/java/com/fox/algorithm/leetcode/链表/_206反转链表.java
package com.fox.algorithm.leetcode.链表;

/**
 * @Author fox
 * @Date 2020/2/18 19:53
 * https://leetcode-cn.com/problems/reverse-linked-list/
 */
public class _206反转链表 {
    //使用递归的方式
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode newHead = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }

    //使用循环的方式
    public ListNode reverseList1(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode newHead = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = newHead;
            newHead = head;
            head = temp;
        }
        return newHead;
    }
}

File: app/src/main/java/com/fox/algorithm/数据结构与算法/leetcode/链表/206反转链表.kt
package com.fox.algorithm.数据结构与算法.leetcode.链表

/**
 * @Author fox
 * @Date 2020/2/18 19:53
 */

/*https://leetcode-cn.com/problems/reverse-linked-list/
反转一个单链表。
示例:
输入: 1->2->3->4->5->NULL
输出: 5->4->3->2->1->NULL
进阶:
你可以迭代或递归地反转链表。你能否用两种方法解决这道题？*/

//使用递归的方式
fun reverseList(head: ListNode?): ListNode? {
    if (head?.next == null) {
        return head
    }
    val newHead = reverseList(head.next)
    head.next!!.next = head
    head.next = null
    return newHead
}

//使用循环的方式
fun reverseList1(head: ListNode?): ListNode? {
    var headTemp = head
    if (headTemp?.next == null) {
        return headTemp
    }
    var newHead: ListNode? = null
    while (headTemp != null) {
        val temp = headTemp.next
        headTemp.next = newHead
        newHead = headTemp
        headTemp = temp
    }
    return newHead
}
